graphql/query_resolver.go

```diff
@@ -982,44 +982,79 @@ func (r *queryResolver) Waterfall(ctx context.Context, options WaterfallOptions)
+
 	opts := model.WaterfallOptions{
 		Limit:      limit,
 		Requesters: requesters,
+		MaxOrder:   maxOrderOpt,
+		MinOrder:   minOrderOpt,
+	}
+
+	activeVersions, err := model.GetActiveWaterfallVersions(ctx, projectId, opts)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting active waterfall versions: %s", err.Error()))
+	}
+
+	// Since GetAllWaterfallVersions uses an inclusive order range ($gte instead of $gt), add 1 to our minimum range
+	minVersionOrder := minOrderOpt + 1
+	if minOrderOpt == 0 {
+		minVersionOrder = activeVersions[len(activeVersions)-1].RevisionOrderNumber
+	}
+
+	// Same as above, but subtract for max order
+	maxVersionOrder := maxOrderOpt - 1
+	if maxOrderOpt == 0 {
+		maxVersionOrder = activeVersions[0].RevisionOrderNumber
 	}
 
-	versions, err := model.GetWaterfallVersions(ctx, projectId, opts)
+	allVersions, err := model.GetAllWaterfallVersions(ctx, projectId, minVersionOrder, maxVersionOrder)
 	if err != nil {
 		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting waterfall versions: %s", err.Error()))
 	}
 
 	// TODO DEVPROD-10179: Add check to ensure each version has tasks that match filter...
 	// Something like this: https://github.com/evergreen-ci/evergreen/blob/bf8f12ec2eefe61f0cf9bcc594924c7be8f91d1b/graphql/query_resolver.go#L869-L938
-	// We should also be able to add a check to the version's build_variant_status field in case of a build variant filter (TODO DEVPROD-10178).
-	// For now, without filters it is guaranteed that `limit` matching versions have been returned.
+	// All other filters can be applied in the GetActiveWaterfallVersions pipeline, ensuring `limit` matching versions have been returned.
 
-	// TODO DEVPROD-10177: Implementing pagination will also allow us to fetch more versions if GetWaterfallVersions does not return enough *activated* versions
+	activeVersionIds := []string{}
+	for _, v := range activeVersions {
+		activeVersionIds = append(activeVersionIds, v.Id)
+	}
+
+	waterfallVersions := groupInactiveVersions(ctx, activeVersionIds, allVersions)
 
-	buildVariants, err := model.GetWaterfallBuildVariants(ctx, versions)
+	buildVariants, err := model.GetWaterfallBuildVariants(ctx, activeVersionIds)
 	if err != nil {
 		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting waterfall build variants: %s", err.Error()))
 	}
 
-	apiVersions := []*restModel.APIVersion{}
-	for _, v := range versions {
-		apiVersion := restModel.APIVersion{}
-		apiVersion.BuildFromService(v)
-		apiVersions = append(apiVersions, &apiVersion)
-	}
-
 	bv := []*model.WaterfallBuildVariant{}
 	for _, b := range buildVariants {
 		bCopy := b
 		bv = append(bv, &bCopy)
 	}
 
+	// If a max order was specified, use this as the new min and vice versa. If not, take this value from the bounds of the returned versions.
+	prevPageOrder := allVersions[0].RevisionOrderNumber
+	if maxOrderOpt > prevPageOrder {
+		prevPageOrder = maxOrderOpt
+	}
```

Comment:
what's a scenario in which this would happen? sorry I tried writing it out but couldn't figure it out 😔

Reply:
Yeah this is not a real scenario, good catch Minna!